From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <finlayoxby@outlook.com>
Date: Fri, 3 Sep 2021 23:02:59 +0100
Subject: [PATCH] Iron Golems Attack Creepers


diff --git a/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyWorldConfig.java b/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyWorldConfig.java
index 4d790ca3bd78f5a4a84b0536a31a4a57ed04b788..05fb6c98c28d4a97da8e35c033ebdb4a95e6a762 100644
--- a/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyWorldConfig.java
+++ b/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyWorldConfig.java
@@ -94,4 +94,11 @@ public final class CrunchyWorldConfig {
         this.chickenAvoidFoxFast = getDouble("mobs.chicken.avoid-fox.fast-speed", 1.5D);
     }
 
+    // Iron Golem Settings
+    public boolean ironGolemAttackCreepers;
+
+    private void ironGolemSettings() {
+        this.ironGolemAttackCreepers = getBoolean("mobs.iron-golem.attack-creepers", true);
+    }
+
 }
diff --git a/src/main/java/net/minecraft/world/entity/animal/IronGolem.java b/src/main/java/net/minecraft/world/entity/animal/IronGolem.java
index ec00c2dd8f969eb99ec6a014e3bcd09c7484b237..dd9d074d1bea2eea56e12f443eafe91ea02373f5 100644
--- a/src/main/java/net/minecraft/world/entity/animal/IronGolem.java
+++ b/src/main/java/net/minecraft/world/entity/animal/IronGolem.java
@@ -82,9 +82,17 @@ public class IronGolem extends AbstractGolem implements NeutralMob {
         this.targetSelector.addGoal(1, new DefendVillageTargetGoal(this));
         this.targetSelector.addGoal(2, new HurtByTargetGoal(this, new Class[0]));
         this.targetSelector.addGoal(3, new NearestAttackableTargetGoal<>(this, Player.class, 10, true, false, this::isAngryAt));
-        this.targetSelector.addGoal(3, new NearestAttackableTargetGoal<>(this, Mob.class, 5, false, false, (entityliving) -> {
-            return entityliving instanceof Enemy && !(entityliving instanceof Creeper);
-        }));
+        // Crunchy start - Iron Golems can attack Creepers
+        if (this.level.crunchyConfig.ironGolemAttackCreepers) {
+            this.targetSelector.addGoal(3, new NearestAttackableTargetGoal<>(this, Mob.class, 5, false, false, (entity) -> {
+                return entity instanceof Enemy;
+            }));
+        } else {
+            this.targetSelector.addGoal(3, new NearestAttackableTargetGoal<>(this, Mob.class, 5, false, false, (entity) -> {
+                return entity instanceof Enemy && !(entity instanceof Creeper);
+            }));
+        }
+        // Crunchy end - Iron Golems can attack Creepers
         this.targetSelector.addGoal(4, new ResetUniversalAngerTargetGoal<>(this, false));
     }
 
