From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <finlayoxby@outlook.com>
Date: Wed, 20 Oct 2021 22:11:59 +0100
Subject: [PATCH] Toggle Reload Command


diff --git a/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyConfig.java b/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyConfig.java
index 37f14a3cfde61a7384faf53b9805b4454a558ab1..7459b9239aedf1a0e763e67319c293d5e704f3bb 100644
--- a/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyConfig.java
+++ b/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyConfig.java
@@ -247,4 +247,11 @@ public final class CrunchyConfig {
         }
     }
 
+    // Disable Reload Command
+    public static boolean disableReloadCommand;
+
+    private static void disableReloadCommand() {
+        disableReloadCommand = getBoolean("commands.disable-reload-command", false);
+    }
+
 }
diff --git a/src/main/java/net/minecraft/commands/Commands.java b/src/main/java/net/minecraft/commands/Commands.java
index f6b73f8c6638ddf79e45042f5c8902ea1f271f5c..2b23f19a6c85c68dddef8e43e81539ac7aa25cc9 100644
--- a/src/main/java/net/minecraft/commands/Commands.java
+++ b/src/main/java/net/minecraft/commands/Commands.java
@@ -103,6 +103,7 @@ import net.minecraft.server.commands.WhitelistCommand;
 import net.minecraft.server.commands.WorldBorderCommand;
 import net.minecraft.server.commands.data.DataCommands;
 import net.minecraft.server.level.ServerPlayer;
+import org.apache.logging.log4j.Level; // Crunchy
 import net.minecraft.util.profiling.jfr.JvmProfiler;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
@@ -159,7 +160,13 @@ public class Commands {
         MsgCommand.register(this.dispatcher);
         ParticleCommand.register(this.dispatcher);
         PlaySoundCommand.register(this.dispatcher);
-        ReloadCommand.register(this.dispatcher);
+        // Crunchy start - Toggle reload command
+        if (dev.mrfishcakes.crunchy.config.CrunchyConfig.disableReloadCommand) {
+            LOGGER.log(Level.WARN, "The /reload command is disabled by Crunchy. It will not register nor work!");
+        } else {
+            ReloadCommand.register(this.dispatcher);
+        }
+        // Crunchy end - Toggle reload command
         RecipeCommand.register(this.dispatcher);
         SayCommand.register(this.dispatcher);
         ScheduleCommand.register(this.dispatcher);
