From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <finlayoxby@outlook.com>
Date: Thu, 2 Sep 2021 16:13:56 +0100
Subject: [PATCH] Crunchy Branding


diff --git a/build.gradle.kts b/build.gradle.kts
index 241ee5c98275a4517e040c86355ae6702f8efca1..445078ff55c537cc041d4f30c3d701d915199d0d 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -1,3 +1,5 @@
+import java.util.Locale
+
 plugins {
     `java-library`
     `maven-publish`
@@ -40,6 +42,9 @@ dependencies {
     apiAndDocs("net.kyori:adventure-text-serializer-plain")
     api("org.apache.logging.log4j:log4j-api:2.14.1")
     api("org.slf4j:slf4j-api:1.8.0-beta4")
+    api("com.zaxxer:HikariCP:5.0.0") // Crunchy
+    api("net.kyori:adventure-text-minimessage:4.1.0-SNAPSHOT") // Crunchy
+    api("com.h2database:h2:2.0.202") // Crunchy
 
     implementation("org.ow2.asm:asm:9.2")
     implementation("org.ow2.asm:asm-commons:9.2")
@@ -67,6 +72,7 @@ dependencies {
 
 configure<PublishingExtension> {
     publications.create<MavenPublication>("maven") {
+        artifactId = project.name.toLowerCase(Locale.ENGLISH)
         from(components["java"])
     }
 }
@@ -113,5 +119,7 @@ tasks.withType<Javadoc> {
         "https://jd.adventure.kyori.net/text-serializer-legacy/$adventureVersion/",
         "https://jd.adventure.kyori.net/text-serializer-plain/$adventureVersion/",
         // Paper end
+        "https://javadoc.io/doc/com.zaxxer/HikariCP/latest/index.html", // Crunchy
+        "https://javadoc.io/doc/com.h2database/h2/latest/index.html", // Crunchy
     )
 }
diff --git a/src/main/java/dev/mrfishcakes/crunchy/CrunchyUtil.java b/src/main/java/dev/mrfishcakes/crunchy/CrunchyUtil.java
new file mode 100644
index 0000000000000000000000000000000000000000..6eb404f228c7353c2a4b02d979f85eb6fa3ba54f
--- /dev/null
+++ b/src/main/java/dev/mrfishcakes/crunchy/CrunchyUtil.java
@@ -0,0 +1,33 @@
+package dev.mrfishcakes.crunchy;
+
+import net.kyori.adventure.text.minimessage.MiniMessage;
+import org.bukkit.ChatColor;
+import org.bukkit.command.CommandSender;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Map;
+
+/**
+ * Class for managing simple/repetitive methods used within Crunchy
+ *
+ * @author MrFishCakes
+ */
+public final class CrunchyUtil {
+
+    private CrunchyUtil() {
+        throw new UnsupportedOperationException("CrunchyUtil cannot be initiated");
+    }
+
+    public static void sendMessage(@NotNull CommandSender sender, @NotNull String message) {
+        sender.sendMessage(ChatColor.translateAlternateColorCodes('&', message.replaceAll("''", "'")));
+    }
+
+    public static void sendMessage(@NotNull MiniMessage miniMessage, @NotNull CommandSender sender, @NotNull String message) {
+        sender.sendMessage(miniMessage.parse(message.replaceAll("''", "'")));
+    }
+
+    public static void sendMessage(@NotNull MiniMessage miniMessage, @NotNull Map<String, String> placeholders, @NotNull CommandSender sender, @NotNull String message) {
+        sender.sendMessage(miniMessage.parse(message.replaceAll("''", "'"), placeholders));
+    }
+
+}
