From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <finlayoxby@outlook.com>
Date: Fri, 3 Sep 2021 18:41:27 +0100
Subject: [PATCH] Crunchy Config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 6996a61534bbdd34729f1dcd988ae2666bfad2c4..e863599539b84980bdefd8ba8d59e66d062b3c28 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1743,6 +1743,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Paper end
 
+        // Crunchy start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getCrunchyConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Crunchy end
+
         /**
          * Sends the component to the player
          *
