From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <finlayoxby@outlook.com>
Date: Fri, 3 Sep 2021 18:41:27 +0100
Subject: [PATCH] Crunchy Config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 7a3a2ff605e06e42ba2a0263e2c50eed5c24f8e9..d9b2420459529a48480789603cb8db534c37665c 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1899,6 +1899,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Paper end
 
+        // Crunchy start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getCrunchyConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Crunchy end
+
         /**
          * Sends the component to the player
          *
