From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <finlayoxby@outlook.com>
Date: Thu, 12 Aug 2021 16:01:49 +0100
Subject: [PATCH] Crunchy Server Config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 83870d54a5ff1017ad7455e9c931fdee54354434..6fe53add360c8965866a26cf45a613227a381241 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1679,6 +1679,22 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Paper end
 
+        // Tuinity start - add config to timings report
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getTuinityConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Tuinity end - add config to timings report
+
+        // Crunchy start - Server config
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getCrunchyConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Crunchy end - Server config
+
         /**
          * Sends the component to the player
          *
