From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <finlayoxby@outlook.com>
Date: Sat, 11 Sep 2021 22:16:51 +0100
Subject: [PATCH] Uptime API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index b7c45aa5d5f79fd16e59401acfa9c2bed0a5d78c..f4b304684a7e37bd44476e0e5b01eb2c9ccce057 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2134,6 +2134,37 @@ public final class Bukkit {
     }
     // Paper end
 
+    // Crunchy start - Uptime API
+    /**
+     * Get the server uptime in milliseconds
+     *
+     * @return Uptime in milliseconds
+     * @deprecated Use alternative methods for more definitive returns
+     */
+    @Deprecated
+    public static long getUptime() {
+        return server.getUptime();
+    }
+
+    /**
+     * Get the elapsed server uptime in milliseconds
+     *
+     * @return Elapsed uptime in milliseconds
+     */
+    public static long getUptimeElapsed() {
+        return server.getUptimeElapsed();
+    }
+
+    /**
+     * Get the uptime start in milliseconds
+     *
+     * @return Uptime start in milliseconds
+     */
+    public static long getUptimeStart() {
+        return server.getUptimeStart();
+    }
+    // Crunchy end - Uptime API
+
     @NotNull
     public static Server.Spigot spigot() {
         return server.spigot();
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index e863599539b84980bdefd8ba8d59e66d062b3c28..486941b322d97be877b51f3bf727f8bc42bb8a16 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1861,4 +1861,31 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @NotNull
     io.papermc.paper.datapack.DatapackManager getDatapackManager();
     // Paper end
+
+    // Crunchy start - Uptime API
+    /**
+     * Get the server uptime in milliseconds
+     *
+     * @return Uptime in milliseconds
+     * @deprecated Use alternative methods for more definitive returns
+     */
+    @Deprecated
+    default long getUptime() {
+        return getUptimeElapsed();
+    }
+
+    /**
+     * Get the elapsed server uptime in milliseconds
+     *
+     * @return Elapsed uptime in milliseconds
+     */
+    long getUptimeElapsed();
+
+    /**
+     * Get the uptime start in milliseconds
+     *
+     * @return Uptime start in milliseconds
+     */
+    long getUptimeStart();
+    // Crunchy end - Uptime API
 }
