From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <finlayoxby@outlook.com>
Date: Sat, 11 Sep 2021 22:16:51 +0100
Subject: [PATCH] Uptime API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index dcfd22862c79ae01ca7707d9abf0a71cc0d4ad9d..f0c9ba11caf90225e4071da36e89fe27e65a1ecc 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2330,6 +2330,36 @@ public final class Bukkit {
     }
     // Paper end
 
+    // Crunchy start - Uptime API
+    /**
+     * Get the server uptime in milliseconds
+     *
+     * @return Uptime in milliseconds
+     * @deprecated Use alternative methods for more definitive returns
+     */
+    @Deprecated
+    public static long getUptime() {
+        return server.getUptime();
+    }
+
+    /**
+     * Get the elapsed server uptime in milliseconds
+     *
+     * @return Elapsed uptime in milliseconds
+     */
+    public static long getUptimeElapsed() {
+        return server.getUptimeElapsed();
+    }
+
+    /**
+     * Get the uptime start in milliseconds
+     *
+     * @return Uptime start in milliseconds
+     */
+    public static long getUptimeStart() {
+        return server.getUptimeStart();
+    }
+    // Crunchy end - Uptime API
     @NotNull
     public static Server.Spigot spigot() {
         return server.spigot();
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index d9b2420459529a48480789603cb8db534c37665c..95ff3eec18144f383dffce0664d6b4d924eaec49 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2034,4 +2034,30 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     @NotNull org.bukkit.potion.PotionBrewer getPotionBrewer();
     // Paper end
+    // Crunchy start - Uptime API
+    /**
+     * Get the server uptime in milliseconds
+     *
+     * @return Uptime in milliseconds
+     * @deprecated Use alternative methods for more definitive returns
+     */
+    @Deprecated
+    default long getUptime() {
+        return getUptimeElapsed();
+    }
+
+    /**
+     * Get the elapsed server uptime in milliseconds
+     *
+     * @return Elapsed uptime in milliseconds
+     */
+    long getUptimeElapsed();
+
+    /**
+     * Get the uptime start in milliseconds
+     *
+     * @return Uptime start in milliseconds
+     */
+    long getUptimeStart();
+    // Crunchy end - Uptime API
 }
