From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <finlayoxby@outlook.com>
Date: Sat, 11 Sep 2021 22:16:51 +0100
Subject: [PATCH] Uptime API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 53c4e5ca208ee17c7c244e416c537c7b63edf194..b7c993e2b5f541b0ff107f1105c12a09fda69163 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2177,6 +2177,37 @@ public final class Bukkit {
     }
     // Paper end
 
+    // Crunchy start - Uptime API
+    /**
+     * Get the server uptime in milliseconds
+     *
+     * @return Uptime in milliseconds
+     * @deprecated Use alternative methods for more definitive returns
+     */
+    @Deprecated
+    public static long getUptime() {
+        return server.getUptime();
+    }
+
+    /**
+     * Get the elapsed server uptime in milliseconds
+     *
+     * @return Elapsed uptime in milliseconds
+     */
+    public static long getUptimeElapsed() {
+        return server.getUptimeElapsed();
+    }
+
+    /**
+     * Get the uptime start in milliseconds
+     *
+     * @return Uptime start in milliseconds
+     */
+    public static long getUptimeStart() {
+        return server.getUptimeStart();
+    }
+    // Crunchy end - Uptime API
+
     @NotNull
     public static Server.Spigot spigot() {
         return server.spigot();
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index d24be16f605276bcb4a0e522c7225d7c884b8f06..c4751d8aa55ed132bc4c192f108bba40b45813f6 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1896,4 +1896,31 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @NotNull
     io.papermc.paper.datapack.DatapackManager getDatapackManager();
     // Paper end
+
+    // Crunchy start - Uptime API
+    /**
+     * Get the server uptime in milliseconds
+     *
+     * @return Uptime in milliseconds
+     * @deprecated Use alternative methods for more definitive returns
+     */
+    @Deprecated
+    default long getUptime() {
+        return getUptimeElapsed();
+    }
+
+    /**
+     * Get the elapsed server uptime in milliseconds
+     *
+     * @return Elapsed uptime in milliseconds
+     */
+    long getUptimeElapsed();
+
+    /**
+     * Get the uptime start in milliseconds
+     *
+     * @return Uptime start in milliseconds
+     */
+    long getUptimeStart();
+    // Crunchy end - Uptime API
 }
