From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <finlayoxby@outlook.com>
Date: Sat, 11 Sep 2021 22:16:51 +0100
Subject: [PATCH] Uptime API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 944f9b87a11472ac6d7e328acc00bf09f899e648..8eef858d67d61221554894ea032aa9e68295f03d 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2308,6 +2308,37 @@ public final class Bukkit {
     }
     // Paper end
 
+    // Crunchy start - Uptime API
+    /**
+     * Get the server uptime in milliseconds
+     *
+     * @return Uptime in milliseconds
+     * @deprecated Use alternative methods for more definitive returns
+     */
+    @Deprecated
+    public static long getUptime() {
+        return server.getUptime();
+    }
+
+    /**
+     * Get the elapsed server uptime in milliseconds
+     *
+     * @return Elapsed uptime in milliseconds
+     */
+    public static long getUptimeElapsed() {
+        return server.getUptimeElapsed();
+    }
+
+    /**
+     * Get the uptime start in milliseconds
+     *
+     * @return Uptime start in milliseconds
+     */
+    public static long getUptimeStart() {
+        return server.getUptimeStart();
+    }
+    // Crunchy end - Uptime API
+
     @NotNull
     public static Server.Spigot spigot() {
         return server.spigot();
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index a354dc32daa78f5f28f990ac58a8df2fb924c979..f7cdd04f961af81b36570e206d97f3baab62e508 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2011,4 +2011,31 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     @NotNull org.bukkit.potion.PotionBrewer getPotionBrewer();
     // Paper end
+
+    // Crunchy start - Uptime API
+    /**
+     * Get the server uptime in milliseconds
+     *
+     * @return Uptime in milliseconds
+     * @deprecated Use alternative methods for more definitive returns
+     */
+    @Deprecated
+    default long getUptime() {
+        return getUptimeElapsed();
+    }
+
+    /**
+     * Get the elapsed server uptime in milliseconds
+     *
+     * @return Elapsed uptime in milliseconds
+     */
+    long getUptimeElapsed();
+
+    /**
+     * Get the uptime start in milliseconds
+     *
+     * @return Uptime start in milliseconds
+     */
+    long getUptimeStart();
+    // Crunchy end - Uptime API
 }
