From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <finlayoxby@outlook.com>
Date: Sat, 11 Sep 2021 22:16:51 +0100
Subject: [PATCH] Uptime API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index d36e4bc3c3713407704b865574cba28662f17315..9862f105a13bdfb7b05988db148a8a3751124a97 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2059,6 +2059,17 @@ public final class Bukkit {
     }
     // Paper end
 
+    // Crunchy start - Uptime API
+    /**
+     * Get the server uptime in milliseconds
+     *
+     * @return Uptime in milliseconds
+     */
+    public static long getUptime() {
+        return server.getUptime();
+    }
+    // Crunchy start - Uptime API
+
     @NotNull
     public static Server.Spigot spigot() {
         return server.spigot();
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 83fa8d2d901898c20e9a3c8018feeeaff0ee6b57..8c563d20e0ef6fa28878518dd02296d63b2bc06d 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1797,4 +1797,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @NotNull
     io.papermc.paper.datapack.DatapackManager getDatapackManager();
     // Paper end
+
+    // Crunchy start - Uptime API
+    /**
+     * Get the server uptime in milliseconds
+     *
+     * @return Uptime in milliseconds
+     */
+    long getUptime();
+    // Crunchy end - Uptime API
 }
