From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <finlayoxby@outlook.com>
Date: Sat, 11 Sep 2021 22:16:51 +0100
Subject: [PATCH] Uptime API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index c093ba024cfd951be1afc4100d899ed8f70ab003..f33eef6e0a4c158cb088859bc9eaec771ad970ec 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2318,6 +2318,36 @@ public final class Bukkit {
     }
     // Paper end
 
+    // Crunchy start - Uptime API
+    /**
+     * Get the server uptime in milliseconds
+     *
+     * @return Uptime in milliseconds
+     * @deprecated Use alternative methods for more definitive returns
+     */
+    @Deprecated
+    public static long getUptime() {
+        return server.getUptime();
+    }
+
+    /**
+     * Get the elapsed server uptime in milliseconds
+     *
+     * @return Elapsed uptime in milliseconds
+     */
+    public static long getUptimeElapsed() {
+        return server.getUptimeElapsed();
+    }
+
+    /**
+     * Get the uptime start in milliseconds
+     *
+     * @return Uptime start in milliseconds
+     */
+    public static long getUptimeStart() {
+        return server.getUptimeStart();
+    }
+    // Crunchy end - Uptime API
     @NotNull
     public static Server.Spigot spigot() {
         return server.spigot();
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 25627a90efbb18bffe8dfeb1c6e0f583786ba732..8f48aacf8ca06a2db986a0e9f9781fbdc5b3de3e 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2021,4 +2021,30 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     @NotNull org.bukkit.potion.PotionBrewer getPotionBrewer();
     // Paper end
+    // Crunchy start - Uptime API
+    /**
+     * Get the server uptime in milliseconds
+     *
+     * @return Uptime in milliseconds
+     * @deprecated Use alternative methods for more definitive returns
+     */
+    @Deprecated
+    default long getUptime() {
+        return getUptimeElapsed();
+    }
+
+    /**
+     * Get the elapsed server uptime in milliseconds
+     *
+     * @return Elapsed uptime in milliseconds
+     */
+    long getUptimeElapsed();
+
+    /**
+     * Get the uptime start in milliseconds
+     *
+     * @return Uptime start in milliseconds
+     */
+    long getUptimeStart();
+    // Crunchy end - Uptime API
 }
