From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <finlayoxby@outlook.com>
Date: Tue, 24 Aug 2021 23:10:16 +0100
Subject: [PATCH] Option to disable withering from Wither Skeletons


diff --git a/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyWorldConfig.java b/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyWorldConfig.java
index 9a02daf3a384145f7a139d7f70dd8ac606fa8992..6ebe927d7924f85e801411c5795954996839382b 100644
--- a/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyWorldConfig.java
+++ b/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyWorldConfig.java
@@ -74,4 +74,12 @@ public final class CrunchyWorldConfig {
         pistonPushDistance = getInt("gameplay.mechanics.piston-push-distance", 12);
     }
 
+    // Disable withering from skeleton
+
+    public boolean disableWitherSkeletonWithering;
+
+    private void witherSkeletonSettings() {
+        disableWitherSkeletonWithering = getBoolean("gameplay.mobs.wither-skeleton.disable-withering", false);
+    }
+
 }
diff --git a/src/main/java/net/minecraft/world/entity/monster/WitherSkeleton.java b/src/main/java/net/minecraft/world/entity/monster/WitherSkeleton.java
index f9bc07be70fae9aced51a69bccb5eda309187a47..86ad03b8764073f13c04964c9fdfc5f12e2a23f2 100644
--- a/src/main/java/net/minecraft/world/entity/monster/WitherSkeleton.java
+++ b/src/main/java/net/minecraft/world/entity/monster/WitherSkeleton.java
@@ -105,6 +105,7 @@ public class WitherSkeleton extends AbstractSkeleton {
             return false;
         } else {
             if (target instanceof LivingEntity) {
+                if (this.level.crunchyConfig.disableWitherSkeletonWithering) return false; // Crunchy - Option to disable withering from skeletons
                 ((LivingEntity) target).addEffect(new MobEffectInstance(MobEffects.WITHER, 200), this, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); // CraftBukkit
             }
 
