From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <finlayoxby@outlook.com>
Date: Sat, 11 Sep 2021 22:16:26 +0100
Subject: [PATCH] Uptime API


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index e98a5aa7b0c61af05d9752b6ce10a4ffa9c0a270..7569360006c46a5b62b9b228e7c388dc947a8bda 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -310,6 +310,8 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     public volatile Thread shutdownThread; // Paper
     public volatile boolean abnormalExit = false; // Paper
 
+    public static final long START_MILLIS = System.currentTimeMillis(); // Crunchy - Uptime API
+
     public static <S extends MinecraftServer> S spin(Function<Thread, S> serverFactory) {
         AtomicReference<S> atomicreference = new AtomicReference();
         Thread thread = new Thread(() -> {
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index d7fca4eb14a8046d807ba743bd6b5a8ddfb3805d..0825b3020b70a36d7471a1953a492d2c1e910ae1 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -2714,4 +2714,16 @@ public final class CraftServer implements Server {
     }
 
     // Paper end
+
+    // Crunchy start - Uptime API
+    @Override
+    public long getUptimeElapsed() {
+        return System.currentTimeMillis() - MinecraftServer.START_MILLIS;
+    }
+
+    @Override
+    public long getUptimeStart() {
+        return MinecraftServer.START_MILLIS;
+    }
+    // Crunchy end - Uptime API
 }
