From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <finlayoxby@outlook.com>
Date: Sat, 11 Sep 2021 22:16:26 +0100
Subject: [PATCH] Uptime API


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 902c7b0cba49cb581d93dc1d70ec7e88bf2e36a1..ef46ddd04354de6e179eff7f74bd691da63a4fee 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -311,6 +311,8 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     public volatile Thread shutdownThread; // Paper
     public volatile boolean abnormalExit = false; // Paper
 
+    public static final long START_MILLIS = System.currentTimeMillis(); // Crunchy - Uptime API
+
     public static <S extends MinecraftServer> S spin(Function<Thread, S> serverFactory) {
         AtomicReference<S> atomicreference = new AtomicReference();
         Thread thread = new Thread(() -> {
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 9778081c9d0fc3238712d20090660c67ecc0678b..5ff0203a63c61380dbe58d1aa6aeed8e9804d674 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -2833,4 +2833,16 @@ public final class CraftServer implements Server {
     }
 
     // Paper end
+
+    // Crunchy start - Uptime API
+    @Override
+    public long getUptimeElapsed() {
+        return System.currentTimeMillis() - MinecraftServer.START_MILLIS;
+    }
+
+    @Override
+    public long getUptimeStart() {
+        return MinecraftServer.START_MILLIS;
+    }
+    // Crunchy end - Uptime API
 }
