From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <finlayoxby@outlook.com>
Date: Sat, 11 Sep 2021 22:16:26 +0100
Subject: [PATCH] Uptime API


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 92e3128cf53b39969d93043d02a49e0feb2a4dd8..9587902be86c44c861b08a45a4d90c5d248c6adb 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -311,6 +311,8 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     public volatile Thread shutdownThread; // Paper
     public volatile boolean abnormalExit = false; // Paper
 
+    public static final long START_MILLIS = System.currentTimeMillis(); // Crunchy - Uptime API
+
     public static <S extends MinecraftServer> S spin(Function<Thread, S> serverFactory) {
         AtomicReference<S> atomicreference = new AtomicReference();
         Thread thread = new Thread(() -> {
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 5d0bff7fa0e4c94f6f61d4ae1630c7d05bd7c6ed..dfbd85634a359cb0bd696b70eac60db3ad6b34e6 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -2805,4 +2805,16 @@ public final class CraftServer implements Server {
     }
 
     // Paper end
+
+    // Crunchy start - Uptime API
+    @Override
+    public long getUptimeElapsed() {
+        return System.currentTimeMillis() - MinecraftServer.START_MILLIS;
+    }
+
+    @Override
+    public long getUptimeStart() {
+        return MinecraftServer.START_MILLIS;
+    }
+    // Crunchy end - Uptime API
 }
