From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <finlayoxby@outlook.com>
Date: Sat, 11 Sep 2021 22:16:26 +0100
Subject: [PATCH] Uptime API


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 503b124f3b3b1c5bbd84ff96f4d2fba30b662076..b2c6c1b5f737c28bf1f2c4c7d2c366c28aa3d67a 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -313,6 +313,8 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     public volatile Thread shutdownThread; // Paper
     public volatile boolean abnormalExit = false; // Paper
 
+    public static final long START_MILLIS = System.currentTimeMillis(); // Crunchy - Uptime API
+
     public static <S extends MinecraftServer> S spin(Function<Thread, S> serverFactory) {
         AtomicReference<S> atomicreference = new AtomicReference();
         Thread thread = new Thread(() -> {
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 7aee1c72b807b1f653f5ac52259c46cf7d14b468..507337b6891ffc8e79d32206a92f9fcdb9b9911f 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -2713,4 +2713,16 @@ public final class CraftServer implements Server {
     }
 
     // Paper end
+
+    // Crunchy start - Uptime API
+    @Override
+    public long getUptimeElapsed() {
+        return System.currentTimeMillis() - MinecraftServer.START_MILLIS;
+    }
+
+    @Override
+    public long getUptimeStart() {
+        return MinecraftServer.START_MILLIS;
+    }
+    // Crunchy end - Uptime API
 }
