From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <finlayoxby@outlook.com>
Date: Thu, 26 Aug 2021 17:03:21 +0100
Subject: [PATCH] Modify Baby Zombie Speed


diff --git a/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyWorldConfig.java b/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyWorldConfig.java
index 0de801f9b6d3b6e0023257d86a2f8a85d00fd423..cdafa47d9fc7d3fb7494e20b638e7b5c5a08b7ca 100644
--- a/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyWorldConfig.java
+++ b/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyWorldConfig.java
@@ -82,6 +82,13 @@ public final class CrunchyWorldConfig {
         disableWitherSkeletonWithering = getBoolean("gameplay.mobs.wither-skeleton.disable-withering", false);
     }
 
+    // Zombie settings
+    public double babyZombieSpeedModifier;
+
+    private void zombieSettings() {
+        babyZombieSpeedModifier = getDouble("gameplay.mobs.zombie.baby-speed-modifier", 0.5D);
+    }
+
     // GameMode settings
     public boolean creativeInstantKill;
 
diff --git a/src/main/java/net/minecraft/world/entity/monster/Zombie.java b/src/main/java/net/minecraft/world/entity/monster/Zombie.java
index 6408e158745c20ab449c44a28420bc9b28e1efac..50fc776cccedebb9b6ea040d61911c3c3554bd95 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Zombie.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Zombie.java
@@ -79,7 +79,8 @@ import org.bukkit.event.entity.EntityTransformEvent;
 public class Zombie extends Monster {
 
     private static final UUID SPEED_MODIFIER_BABY_UUID = UUID.fromString("B9766B59-9566-4402-BC1F-2EE2A276D836");
-    private final AttributeModifier SPEED_MODIFIER_BABY = new AttributeModifier(Zombie.SPEED_MODIFIER_BABY_UUID, "Baby speed boost", 0.5D, AttributeModifier.Operation.MULTIPLY_BASE); private final AttributeModifier babyModifier = this.SPEED_MODIFIER_BABY; // Paper - remove static - Make baby speed configurable
+    private final AttributeModifier SPEED_MODIFIER_BABY = new AttributeModifier(Zombie.SPEED_MODIFIER_BABY_UUID, "Baby speed boost", this.level.crunchyConfig.babyZombieSpeedModifier, AttributeModifier.Operation.MULTIPLY_BASE); // Crunchy - Modify baby zombie speed
+    private final AttributeModifier babyModifier = this.SPEED_MODIFIER_BABY; // Crunchy // Paper - remove static - Make baby speed configurable
     private static final EntityDataAccessor<Boolean> DATA_BABY_ID = SynchedEntityData.defineId(Zombie.class, EntityDataSerializers.BOOLEAN);
     private static final EntityDataAccessor<Integer> DATA_SPECIAL_TYPE_ID = SynchedEntityData.defineId(Zombie.class, EntityDataSerializers.INT);
     public static final EntityDataAccessor<Boolean> DATA_DROWNED_CONVERSION_ID = SynchedEntityData.defineId(Zombie.class, EntityDataSerializers.BOOLEAN);
