From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <finlayoxby@outlook.com>
Date: Sun, 7 Nov 2021 18:05:14 +0000
Subject: [PATCH] Dragon Head Acts as Pumpkin


diff --git a/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyWorldConfig.java b/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyWorldConfig.java
index 077ed729b1339f7d2694710a8685ec32716fd095..7127455c0527e42abf5ce0b91927fe4aa7fe9c12 100644
--- a/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyWorldConfig.java
+++ b/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyWorldConfig.java
@@ -103,6 +103,8 @@ public final class CrunchyWorldConfig {
     public String nuclearCreeperName;
     public boolean nuclearCreeperShowName;
 
+    public boolean dragonHeadBypass;
+
     private void mobSettings() {
         this.nuclearCreeperEnabled = getBoolean("mobs.creeper.nuclear.enabled", false);
         this.nuclearCreeperChanceEasy = getDouble("mobs.creeper.nuclear.easy-spawn-chance", 0.0D);
@@ -111,6 +113,8 @@ public final class CrunchyWorldConfig {
         this.nuclearCreeperPower = (float) getDouble("mobs.creeper.nuclear.power", 3.5D);
         this.nuclearCreeperName = getString("mobs.creeper.nuclear.name", "<#2cfa1f>Nuclear Creeper");
         this.nuclearCreeperShowName = getBoolean("mobs.creeper.nuclear.show-name", true);
+
+        this.dragonHeadBypass = getBoolean("mobs.enderman.dragon-head-bypass", false);
     }
 
     // World Settings
diff --git a/src/main/java/net/minecraft/world/entity/monster/EnderMan.java b/src/main/java/net/minecraft/world/entity/monster/EnderMan.java
index 4920f5465298b4da33b53f80b8af6220cb0e8ccc..6a14554cc9679860923e3079a008fa244eaf69b6 100644
--- a/src/main/java/net/minecraft/world/entity/monster/EnderMan.java
+++ b/src/main/java/net/minecraft/world/entity/monster/EnderMan.java
@@ -236,7 +236,7 @@ public class EnderMan extends Monster implements NeutralMob {
         // Paper end
         ItemStack itemstack = (ItemStack) player.getInventory().armor.get(3);
 
-        if (itemstack.is(Blocks.CARVED_PUMPKIN.asItem())) {
+        if (itemstack.is(Blocks.CARVED_PUMPKIN.asItem()) || (this.level.crunchyConfig.dragonHeadBypass && itemstack.is(net.minecraft.world.item.Items.DRAGON_HEAD))) { // Crunchy - Dragon head acts as pumpkin
             return false;
         } else {
             Vec3 vec3d = player.getViewVector(1.0F).normalize();
