From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <finlayoxby@outlook.com>
Date: Wed, 20 Oct 2021 22:11:59 +0100
Subject: [PATCH] Toggle Reload Command


diff --git a/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyConfig.java b/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyConfig.java
index 69156dba8fc49154180ef288eb39fee582bbec6d..767421b4de69e87bb0eb3be4c278b68a614b80ff 100644
--- a/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyConfig.java
+++ b/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyConfig.java
@@ -272,4 +272,11 @@ public final class CrunchyConfig {
         }
     }
 
+    // Disable Reload Command
+    public static boolean disableReloadCommand;
+
+    private static void disableReloadCommand() {
+        disableReloadCommand = getBoolean("commands.disable-reload-command", false);
+    }
+
 }
diff --git a/src/main/java/net/minecraft/commands/Commands.java b/src/main/java/net/minecraft/commands/Commands.java
index edc202d60ac3e8ac13b61d8c2c82d587567e9c05..007e9cfbfb0f5db5e91afa5ee3aeca165dfbbbe6 100644
--- a/src/main/java/net/minecraft/commands/Commands.java
+++ b/src/main/java/net/minecraft/commands/Commands.java
@@ -103,6 +103,7 @@ import net.minecraft.server.commands.WhitelistCommand;
 import net.minecraft.server.commands.WorldBorderCommand;
 import net.minecraft.server.commands.data.DataCommands;
 import net.minecraft.server.level.ServerPlayer;
+import org.apache.logging.log4j.Level; // Crunchy
 import net.minecraft.util.profiling.jfr.JvmProfiler;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
@@ -159,7 +160,13 @@ public class Commands {
         MsgCommand.register(this.dispatcher);
         ParticleCommand.register(this.dispatcher);
         PlaySoundCommand.register(this.dispatcher);
-        ReloadCommand.register(this.dispatcher);
+        // Crunchy start - Toggle reload command
+        if (dev.mrfishcakes.crunchy.config.CrunchyConfig.disableReloadCommand) {
+            LOGGER.log(Level.WARN, "The /reload command is disabled by Crunchy. It will not register nor work!");
+        } else {
+            ReloadCommand.register(this.dispatcher);
+        }
+        // Crunchy end - Toggle reload command
         RecipeCommand.register(this.dispatcher);
         SayCommand.register(this.dispatcher);
         ScheduleCommand.register(this.dispatcher);
