From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <fishcake007@outlook.com>
Date: Wed, 15 Dec 2021 22:40:35 +0000
Subject: [PATCH] Enhanced Plugins Command


diff --git a/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyConfig.java b/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyConfig.java
index 5e32de77a0615123858551870f27a454d6a2cdb4..793e708662f715a4ec686641c351c35b678805e2 100644
--- a/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyConfig.java
+++ b/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyConfig.java
@@ -174,9 +174,11 @@ public final class CrunchyConfig {
 
     // Command Settings
     public static String uptimeCommandOutput;
+    public static boolean enhancedPluginsCommand;
 
     private static void commandSettings() {
         uptimeCommandOutput = getString("commands.uptime-output", "<white>The server has been up for: <green><days><white> days, <green><hours><white> hours, <green><minutes><white> minutes");
+        enhancedPluginsCommand = getBoolean("commands.enhanced-plugins", true);
     }
 
     // Recipe Settings
