From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <finlayoxby@outlook.com>
Date: Wed, 8 Dec 2021 17:27:07 +0000
Subject: [PATCH] Iron Golems Attack Creepers


diff --git a/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyWorldConfig.java b/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyWorldConfig.java
index cc4fef22a0453b12f20f7c706b527f5e4c0bc302..172efc9a50d6534268fe75c9535d2ff50de745f6 100644
--- a/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyWorldConfig.java
+++ b/src/main/java/dev/mrfishcakes/crunchy/config/CrunchyWorldConfig.java
@@ -134,6 +134,7 @@ public final class CrunchyWorldConfig {
     public float nuclearCreeperPower;
     public String nuclearCreeperName;
     public boolean nuclearCreeperShowName;
+    public boolean ironGolemsAttackCreepers;
 
     private void creeperSettings() {
         this.nuclearCreeper = getBoolean("mobs.creeper.nuclear.enabled", false);
@@ -143,6 +144,8 @@ public final class CrunchyWorldConfig {
         this.nuclearCreeperPower = (float) getDouble("mobs.creeper.nuclear.power", 3.5D);
         this.nuclearCreeperName = getString("mobs.creeper.nuclear.name", "<#2cfa1f>Nuclear Creeper");
         this.nuclearCreeperShowName = getBoolean("mobs.creeper.nuclear.show-name", true);
+
+        this.ironGolemsAttackCreepers = getBoolean("mobs.iron-golem.attack-creepers", false);
     }
 
     // Enderman Settings
diff --git a/src/main/java/net/minecraft/world/entity/animal/IronGolem.java b/src/main/java/net/minecraft/world/entity/animal/IronGolem.java
index d6bff18a60e1b0b507a3797742bfafff2fad10d2..5d1d5982b279aca56f2fef7ba93f29d9b708867c 100644
--- a/src/main/java/net/minecraft/world/entity/animal/IronGolem.java
+++ b/src/main/java/net/minecraft/world/entity/animal/IronGolem.java
@@ -55,6 +55,21 @@ import net.minecraft.world.level.material.Fluids;
 import net.minecraft.world.phys.Vec3;
 
 public class IronGolem extends AbstractGolem implements NeutralMob {
+    // Crunchy start - Iron golems attack Creepers
+    public static class IronGolemCreeperGoals {
+
+        public static NearestAttackableTargetGoal<Mob> minecraft(Mob mob) {
+            return new NearestAttackableTargetGoal<>(mob, Mob.class, 5, false, false, (entity) ->
+                    entity instanceof Enemy && !(entity instanceof Creeper));
+        }
+
+        public static NearestAttackableTargetGoal<Mob> crunchy(Mob mob) {
+            return new NearestAttackableTargetGoal<>(mob, Mob.class, 5, false, false, (entity) ->
+                    entity instanceof Enemy);
+        }
+
+    }
+    // Crunchy end - Iron golems attack Creepers
 
     protected static final EntityDataAccessor<Byte> DATA_FLAGS_ID = SynchedEntityData.defineId(IronGolem.class, EntityDataSerializers.BYTE);
     private static final int IRON_INGOT_HEAL_AMOUNT = 25;
@@ -82,9 +97,7 @@ public class IronGolem extends AbstractGolem implements NeutralMob {
         this.targetSelector.addGoal(1, new DefendVillageTargetGoal(this));
         this.targetSelector.addGoal(2, new HurtByTargetGoal(this, new Class[0]));
         this.targetSelector.addGoal(3, new NearestAttackableTargetGoal<>(this, Player.class, 10, true, false, this::isAngryAt));
-        this.targetSelector.addGoal(3, new NearestAttackableTargetGoal<>(this, Mob.class, 5, false, false, (entityliving) -> {
-            return entityliving instanceof Enemy && !(entityliving instanceof Creeper);
-        }));
+        this.targetSelector.addGoal(3, level.crunchyConfig.ironGolemsAttackCreepers ? IronGolemCreeperGoals.crunchy(this) : IronGolemCreeperGoals.minecraft(this)); // Crunchy - Iron golems attack Creepers
         this.targetSelector.addGoal(4, new ResetUniversalAngerTargetGoal<>(this, false));
     }
 
@@ -142,7 +155,7 @@ public class IronGolem extends AbstractGolem implements NeutralMob {
 
     @Override
     public boolean canAttackType(EntityType<?> type) {
-        return this.isPlayerCreated() && type == EntityType.PLAYER ? false : (type == EntityType.CREEPER ? false : super.canAttackType(type));
+        return this.isPlayerCreated() && type == EntityType.PLAYER ? false : (type == EntityType.CREEPER ? level.crunchyConfig.ironGolemsAttackCreepers : super.canAttackType(type)); // Crunchy - Iron golems attack Creepers
     }
 
     @Override
